wandb/wandb_agent.py

```diff
@@ -1,517 +1,165 @@
-                            )
-                            self._running = False
-                            break
-                    logger.info("Cleaning up finished run: %s", run_id)
-                    del self._run_processes[run_id]
-                    self._last_report_time = None
-                    self._finished += 1
+        self._agent_id = agent["id"]
 
-                if self._count and self._finished >= self._count or not self._running:
-                    self._running = False
-                    continue
+    def check_queue(self):
+        run_status = dict()
+        commands = self._api.agent_heartbeat(self._agent_id, {}, run_status)
+        if not commands:
+            return
+        command = commands[0]
+        job = Job(command)
+        return job
 
-                commands = self._api.agent_heartbeat(agent_id, {}, run_status)
+    def run_job(self, job):
+        run_id = job.run_id
 
-                # TODO: send _server_responses
-                self._server_responses = []
-                for command in commands:
-                    self._server_responses.append(self._process_command(command))
-
-        except KeyboardInterrupt:
-            try:
-                wandb.termlog(
-                    "Ctrl-c pressed. Waiting for runs to end. Press ctrl-c again to terminate them."
-                )
-                for _, run_process in six.iteritems(self._run_processes):
-                    run_process.wait()
-            except KeyboardInterrupt:
-                pass
-        finally:
-            try:
-                if not self._in_jupyter:
-                    wandb.termlog("Terminating and syncing runs. Press ctrl-c to kill.")
-                for _, run_process in six.iteritems(self._run_processes):
-                    try:
-                        run_process.terminate()
-                    except OSError:
-                        pass  # if process is already dead
-                for _, run_process in six.iteritems(self._run_processes):
-                    run_process.wait()
-            except KeyboardInterrupt:
-                wandb.termlog("Killing runs and quitting.")
-                for _, run_process in six.iteritems(self._run_processes):
-                    try:
-                        run_process.kill()
-                    except OSError:
-                        pass  # if process is already dead
-
-    def _process_command(self, command):
-        logger.info(
-            "Agent received command: %s"
-            % (command["type"] if "type" in command else "Unknown")
-        )
-        response = {
-            "id": command.get("id"),
-            "result": None,
-        }
-        try:
-            command_type = command["type"]
-            result = None
-            if command_type == "run":
-                result = self._command_run(command)
-            elif command_type == "stop":
```

Comment:
where are you handling this? you need to have a thread that is doing a heartbeat checking for stop and exit commands